Added Node-Fighter Game example

webocs opened the pull request:
# PR game example made during NodeConfAr 2018
I'm pull requesting the game we made during the conference, on behalf of the full team.

# Synopsis
Node fighter is an infinite scroller where you are the NodeShip, going around the universe destroying other  programming languages.

# PR Recap
+ Added finished and improved game js file 
+ Added index file

# The Workshop team
Artists:
[@NickoArias](https://twitter.com/NickoArias)
[@maxisalazar](https://twitter.com/maxisalazar)

Programmers
[@alan_derenzis](https://github.com/aderenzis)
[@nahue086](https://twitter.com/nahue086)
[@webocs](https://twitter.com/webocs)

webocs committed:
Added Node-Fighter Game example

+ Added game example js file 
+ Added stars background
+ Added index file

webocs committed:
Removed objects

+ Removed all objects
+ Replaced objects with prefixed variables and arrays

webocs committed:
compilable version

webocs committed:
Added compiled version renamed folder

webocs committed:
Added ino file

webocs committed:
Added arduboy compattible version

+ Fixed issue with random
+ Removed string concats
+ Removed comparisons with sprites and replaced with ints

webocs committed:
Updated Ino added hex

flaki commented:
This is super cool!

I'm looking at the diff of the Arduboy version, trying to figure out why you had to create two separate versions. The palette-downmixing is not perfect so giving separate imagas is understandable, the screen-size related stuff is also understandable (although you were upscaling as the Arduboy screen is wider, it should be solvable), but there are also a lot of other changes.

I'm trying to see if there is any way to create an unified version -- as that would be the point of Clouduboy, is to create one file, but compile for and run on different devices... Could you describe me what you had to change compared to the Tiny Arcade version to get this working on the Arduboy?

webocs commented:
Hey!, there are a few considerations to this. I was not aware of the clouduboy to hex compiler, so what I was doing was to compile to c++ using the builder and generate the hexes using ABE. A few things I can do this evening:

+ Take  a few screenshots of how it looks like with several colors and compare that to the ascii drawing
+ Try building using clouduboy hex builder (I can't test the tiny arcade because there's no emulator)

I'll report after that ;)

EDIT Report:
+ I was not able to  build using https://builder.clouduboy.community/ 
+ This is how everything looks using the colored version on the arduboy (next to the ascii representation)

### Intro
![image](https://user-images.githubusercontent.com/2563911/48037505-4f0f4e00-e14b-11e8-93d5-20926f059f38.png)

### PHP
![image](https://i.snag.gy/9k5ZGa.jpg)

### JAVA
![image](https://user-images.githubusercontent.com/2563911/48037535-782fde80-e14b-11e8-9909-0a109d52a600.png)

### CPP
![image](https://i.snag.gy/5wTR3A.jpg)

### NODE SHIP
![image](https://user-images.githubusercontent.com/2563911/48037557-9269bc80-e14b-11e8-8827-6e2b5ef05609.png)

## Conclusion?
IDK if this makes sense, but seems like even numbers are not rendered correctly.. maybe?

webocs commented:
As for your question on what I changed from the clouduboy compatible version:

### Changes for the arduboy version
+ I made all sprites only white, changed al numbers to # and removed the extra colors from the sprite definitions
+ Modified all spots where the random function appeared multiplied by MAX_SOMETHING to %. This is because random in the arduboy does not range from 0 to 1 (did not compile for the tiny arcade so this may be an issue there too)

### Changes to get the C++ compilable by ABE
+ Loaded all sprites inside the setup function
+ Removed all objects and replaced with prefixed arrays and variables
+ Removed high order functions
+ Removed dynamic arrays and every use of push and splice
+ Removed string concats and template literals
+ Removed gfx comparisons

flaki commented:
This is super helpful feedback, thank you so much for taking your time writing it up! :sparkling_heart:

Regarding the images, yes, I believe the current version of the library does a very rudimentary downsampling based on the palette specified, basically takes every even-number-indexed palette entry as "transparent" and odd indexed colors as "opaque/white". This is, as you see, not ideal (you could work around it with adding "padding" colors, but still). The new version of the pixelart editor and the accompanying microcanvas library I'm working on will fix this issue by allowing you to specify your own "masking palette", that is, you will be able to decide which colors remain opaque on the images when downsampled to 1-bit.

On the CPP conversion, again, very helpful feedback! I have been planning to allow for some of the array methods to be transpiled (to expand on the [current support](https://github.com/clouduboy/clouduboy-compiler/tree/master/modules/transforms/array)), and was thinking on allowing basic, stack-allocated structs with at least only numeric properties to be used. I'm super happy you submitted the "original" version, in the commit history, too -- this will allow me to go back and see "what made sense" to use in a situation for you, and how you had to change it to make compilation possible. Maybe with smarter compilation we can get rid of those restrictions. ;)

webocs commented:
Awesome!  This morning I was thinking about some shading patterns when converting to 1 bit, it's impossible to capture the true essence of the colored version but maybe there's a way to convert shades of colors to patterns with + or x or .  ..  Something like this:

https://29a.ch/ditherlicious/

I might find some time to work on something like that this weekend, maybe it can help with the downsampling process. As for the other, I want to read the full code to see if I can PR some of the missing features. Idk if you're looking for help with that yet but if you are I'll be happy to help.
